### PR TITLE
Flow tests with page objects

### LINKED ISSUE #7: Delete Image
#### As a user I want to delete an image.

__Story__:
Our users are awesome at adding images to our web service. Some of our users
add dozens in a single session. Occasionally, however, some users add the wrong
image, and would like a way to remove those images. Please provide our users
with the ability to delete images. Note that at this point we trust our users
not to abuse this feature and delete more images than they should.

__Acceptance criteria__:
- [x] The user is prompted for confirmation prior to deletion. (Use javascript
  confirm() dialog)
- [x] Deleted images are removed from the homepage.
- [x] A flow test using page objects exists for the flow to create and
  subsequently delete an image.

__Dependencies__:
- Image Index

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm
- https://github.com/appfolio/ae_page_objects
- http://martinfowler.com/bliki/PageObject.html
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects
- https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings#javascript-dialogs

__Flow Test Info__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the [ae_page_objects readme](https://github.com/appfolio/ae_page_objects)
which will be an extremely helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Important Notes__:

**Academy session on Page Objects:** We have an AppFolio Engineering Academy session on Selenium and Page Objects. It would be helpful to attend this session before doing this story, but because it can be hard to schedule everyone, here are the slides and a recording of the Apr 12 2019 session:

- Slides: https://docs.google.com/presentation/d/1ytto5-U93VeCTA9krRg8_c8o_5vadpucp9AbVTEasBA/edit#slide=id.g35b2d5d377_2_5
- Video: https://transcripts.gotomeeting.com/#/s/c1ebcf43f5387db5a610e9ccdff697cd3b9596ab70153fe4059f5bee27750d4d

**Firefox version:** Generally you should upgrade your Firefox to the latest version before attempting to use 
AePageObjects, Capybara, and Selenium. However, sometimes swapping out your Firefox version can resolve weird errors.
For example, between November 2016 and mid-2018, a specific older Firefox version,
47.0.1, was required to work properly with AePageObjects, Capybara, and
Selenium. This version comes on AppFolio developer laptop images by default,
but if you've used Firefox, it may have updated to a newer version that won't
work, and you'll need to replace it (or install it side-by-side) with 47.0.1
version from here:

- https://ftp.mozilla.org/pub/firefox/releases/47.0.1/mac/en-US/

To facilitate swapping between Firefox versions, it's useful to have multiple versions in `/Applications/`:
`Firefox-47.app`, `Firefox-59.app`, `Firefox-64.app`, 
and then simply copy/paste the one you want to use, and rename it  `Firefox.app`. 

Problems that you may see with newer Firefox versions may include
 - failing to open a Firefox window for testing
 - open a Firefox window but failing to load any pages into it
 - other strange issues such as pages loading, but failing to open browser modal alert/confirm dialogs.

**Selenium and JavaScript:** When dealing with javascript confirm() dialog in the flow test, this [reference](https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings#javascript-dialogs)
will be helpful.

### PR DESCRIPTION
**Pull request by maiquynhtruong**
Closes #7

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -1,10 +1,11 @@
 class ImagesController < ApplicationController
   def index
     @images = if filtered_image_params.empty?
-                Image.all.order('created_at DESC')
+                Image.all
               else
-                Image.tagged_with(params[:tag], any: true).by_created_date
+                Image.tagged_with(params[:tag], any: true)
               end
+    @images = @images.order('created_at DESC')
```
[the remark this answers is not in the document]
Comment (joshuasbates) [reply app/controllers/images_controller.rb]: Correct me if I'm wrong, I don't think there's a change of behavior here.  The `by_created_date` scope is defined as `order('created_at DESC')`, so it's just a minor refactor.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -24,6 +25,12 @@ def show
     @image = Image.find(params[:id])
   end
 
+  def destroy
+    @image = Image.find(params[:id])
+    @image.destroy
+    redirect_to images_path
+  end
```
[the remark this answers is not in the document]
Comment (joshuasbates) [reply app/controllers/images_controller.rb]: Also, something nice you can do in Ruby controllers is execute a `before_action` command, and limit it to specific actions.  This is nice when you have some common action preconditions (such as needing to find an image in order to proceed).

In this example, you might want to add to the top of the controller:
```
before_action :find_image, only: [:show, :destroy]
```
Then, in the private section of the controller
```
def find_image
  @image = Image.find(params[:id])
  # Question:  What errors might need to be handled?
end
```
Then you can make the `show` action an empty implementation, and remove the find directive from the `destroy` action.

You don't necessarily need to change this, but it is something to be aware of, and you will see often.

File: app/views/images/show.html.erb [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
+  <% end %>
+</div>
 
-<img src="<%= @image.url %>" style="max-width:400px; max-height:400px;">
 
-<p><%= link_to 'Back to homepage', images_path %> </p>
+<img class='js-image-container' src="<%= @image.url %>" style="max-width:400px; max-height:400px;">
+
+<div>
+  <td><%= link_to 'Delete this image', image_path(@image),
```
[the remark this answers is not in the document]
Comment (maiquynhtruong) [reply app/views/images/show.html.erb]: Changed it to `<p>, which was what I intended 😅

File: app/controllers/images_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (maiquynhtruong) [reply app/controllers/images_controller.rb]: Yes and I explained with David about that and showed him your comment from the previous PR

File: app/views/images/index.html.erb [hunk 4]
```diff
@@ -1,19 +1,23 @@
-<h1><%= link_to 'Image link submission', new_image_path %></h1>
+<h1 class='js-new-image-submission'><%= link_to 'Image link submission', new_image_path %></h1>
```
[the remark this answers is not in the document]
Comment (maiquynhtruong) [reply app/views/images/index.html.erb]: I added some flash messages to display when image is created and deleted successfully

File: app/controllers/images_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (maiquynhtruong) [reply app/controllers/images_controller.rb]: @david: I have added the tests to delete in the index page and in the show page
@josh: I'll look into that. Thank you!

File: app/views/images/index.html.erb [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
 
 <h1>Showing all images</h1>
 
-<table>
+<h2><%= link_to 'Clear filtered tags', images_path %></h2>
```
Comment (joshuasbates) [inline app/views/images/index.html.erb]: This link looks inappropriate for this PR -- there's no way to filter yet, is there?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -128,4 +128,27 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest # rubocop:disable M
...
   end
+
+  test 'index page delete image' do
+    image = Image.create!(url: 'http://abc.jpg')
+
+    get images_path
+    assert_response :success
+
+    delete image_path(image)
+    assert_redirected_to images_path
```
Comment (joshuasbates) [inline test/controllers/images_controller_test.rb]: Minor:  It is also typical to check that the expected flash message is displayed after redirect.

File: app/views/images/index.html.erb (hunk 5 above)
Comment (maiquynhtruong) [reply app/views/images/index.html.erb]: I meant to have this in an index page that shows filtered images by tags so user has a way to go back to all images. I should have implemented this for the previous issue but it didn't come to me until Abineet suggested it yesterday. This is minor addition though and I can remove it.

**Pull request merged**